File: iosApp/iosApp/p2p/SwapMessageNotifiable.swift
//
//  SwapMessageNotifiable.swift
//  iosApp
//
//  Created by jimmyt on 8/24/22.
//  Copyright © 2022 orgName. All rights reserved.
//

import Foundation

/**
 A protocol that a structure or class must adopt in order to be notified of swap-related messages by `P2PService`
 */
protocol SwapMessageNotifiable {
    /**
     The function called by `P2PService` in order to notify the structure or class adopting this protocol of a new `TakerInformationMessage`.
     
     - Parameter message: the new `TakerInformationMessage` of which the structure or class adopting this protocol is being notified, and should handle in the implementation of this function.
     */
    func handleTakerInformationMessage(_ message: TakerInformationMessage) throws
    /**
     The function called by `P2PService` in order to notify the structure or class adopting this protocol of a new `MakerInformationMessage`.
     
     - Parameters:
        - message: the new `MakerInformationMessage` of which the structure or class adopting this protocol is being notified, and should handle in the implementation of this function.
        - senderInterfaceID: The interface ID of the message's sender.
        - recipientInterfaceID: The interface ID of the message's intended recipient.
     */
    func handleMakerInformationMessage(_ message: MakerInformationMessage, senderInterfaceID: Data, recipientInterfaceID: Data) throws
}

File: androidApp/src/main/java/com/commuto/interfacemobile/android/p2p/SwapMessageNotifiable.kt
package com.commuto.interfacemobile.android.p2p

import com.commuto.interfacemobile.android.p2p.messages.MakerInformationMessage
import com.commuto.interfacemobile.android.p2p.messages.TakerInformationMessage

/**
 * An interface that a class must implement in order to be notified of swap-related messages by [P2PService].
 */
interface SwapMessageNotifiable {
    /**
     * The method called by [P2PService] in order to notify the class implementing this interface of a new
     * [TakerInformationMessage].
     *
     * @param message The new [TakerInformationMessage] of which the class implementing this interface is being notified
     * and should handle in the implementation of this method.
     */
    suspend fun handleTakerInformationMessage(message: TakerInformationMessage)

    /**
     * The method called by [P2PService] in order to notify the class implementing this interface of a new
     * [MakerInformationMessage].
     *
     * @param message The new [MakerInformationMessage] of which the class implementing this interface is being notified
     * and should handle in the implementation of this method.
     * @param senderInterfaceID The interface ID of the message's sender.
     * @param recipientInterfaceID The interface ID of the message's intended recipient.
     */
    suspend fun handleMakerInformationMessage(
        message: MakerInformationMessage,
        senderInterfaceID: ByteArray,
        recipientInterfaceID: ByteArray
    )
}